### PR TITLE
Added TextStyle Horizontal Alignment

### LINKED ISSUE #1282: Feature request: allow horizontal centering in TextStyle
## Proposed solution

1. Allow [Align.C](https://py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.Align) to be passed as a value for `.l_margin` in [TextStyle](https://py-pdf.github.io/fpdf2/fpdf/fonts.html#fpdf.fonts.TextStyle)
2. When this `.l_margin` is used in [FPDF._use_title_style()](https://github.com/py-pdf/fpdf2/blob/master/fpdf/fpdf.py#L5126), if it is `Align.C`, then horizontally align the text

The PR implementing this should include unit tests using `FPDF.start_section()`

### Discussed in https://github.com/py-pdf/fpdf2/discussions/1280

<div type='discussions-op-text'>

<sup>Originally posted by **Benoite142** October 10, 2024</sup>:

> Hi,

> I've been trying to center a TitleStyle for quite some time now. I've been using a TitleStyle for the `start_section` function used in the TOC placeholder and I am having a lot of trouble trying to center the TitleStyle. I know we can easily center some cell or text with `align='C'` but I found no way of doing that for the TitleStyle. I know we can put a value to the `l_margin` to make it look like it is centered, but I can't really use that since all of my titles differs in length. 

> I also tried aligning everything to the center of the page, but that still doesn't work with TitleStyle and the sections.
Worst case scenario is that I put everything on the left, but I would really want it to be centered.

> If anyone knows how I could do this, please let me know.

> Thank you for you time 😊</div>

### PR DESCRIPTION
**Pull request by visheshdvivedi**
Added horizontal alignment feature for TextStyle.

Fixes #1282  

**Checklist**:

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged 

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

I couldn't find any docstring and a dedicated section as documentation for TextStyle. Please message me if it exists, will update it accordingly

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by andersonhc** [comment]
@visheshdvivedi 

please remove the "generate=True" from the test below:

Run # Ensuring there is no `generate=True` left remaining in calls to assert_pdf_equal:
test/outline/test_outline.py:    assert_pdf_equal(pdf, HERE / "test_start_section_horizontal_alignment.pdf", tmp_path, generate=True)

**Comment by visheshdvivedi** [comment]
Hi @andersonhc,

Added commit to remove `generate=True`

**Comment by andersonhc** [comment]
> Hi @andersonhc,
> 
> Added commit to remove `generate=True`

The test on python 3.9 is failing due to use of pipe to union. [That feature was introduced in Python 3.10](https://peps.python.org/pep-0604/)

Can you replace it by Union so we can keep compatibility with older Python versions?

**Comment by visheshdvivedi** [comment]
Hi @andersonhc,

Replaced pipe operator with `Union` as suggested.

File: fpdf/fonts.py [hunk 1]
```diff
@@ -136,7 +136,15 @@ def __init__(
...
         self.t_margin = t_margin or 0
-        self.l_margin = l_margin or 0
+
+        # added support for 'Align' and 'str' type values for l_margin
+        if isinstance(l_margin, (Align, int)):
+            self.l_margin = l_margin
+        elif isinstance(l_margin, str):
+            self.l_margin = Align.coerce(l_margin)
+        else:
+            self.l_margin = 0
```
Comment (Lucas-C) [inline fpdf/fonts.py:146]:
```suggestion
        self.l_margin = Align.coerce(l_margin)
```
Isn't this one line enough?

File: fpdf/fpdf.py [hunk 2]
```diff
@@ -5085,6 +5085,10 @@ def start_section(self, name, level=0, strict=True):
             if title_style.size_pt is not None:
                 prev_font_size_pt = self.font_size_pt
                 self.font_size_pt = title_style.size_pt
+            # check if l_margin value is of type Align or string
+            align = Align.L
+            if isinstance(title_style.l_margin, (Align, str)):
+                align = title_style.l_margin
```
Comment (Lucas-C) [inline fpdf/fpdf.py:5091]:
```suggestion
            align = title_style.l_margin or Align.L
```
This single line using `or` to set a fallback default value may be enough?

File: test/outline/test_outline.py [hunk 3]
```diff
@@ -65,6 +65,28 @@ def test_incoherent_start_section_hierarchy():
     with pytest.raises(ValueError):
         pdf.start_section("Subtitle", level=2)
 
+def test_start_section_horizontal_alignment(tmp_path):  # issue-1282
```
Comment (Lucas-C) [inline test/outline/test_outline.py:68]: 👍 Nice test!

**Comment by Lucas-C** [comment]
Good job overall @visheshdvivedi 👍

Once you'll have rebased your branch on this repo `master` branch,
you should find this on line 44 in `fpdf/html.py`:
```python
    "title": TextStyle(  # Only rendered if render_title_tag=True
        b_margin=0.4,
        font_size_pt=30,
        t_margin=6,
        # center=True, - Enable this once #1282 is implemented
    ),
```
This comes from PR https://github.com/py-pdf/fpdf2/pull/1298.
Could you replace the commented out `center=True` line by `l_margin=Align.C` now that you have implemented this feature, please?

File: fpdf/fonts.py (hunk 1 above)
Comment (visheshdvivedi) [reply fpdf/fonts.py:146]: Hi @Lucas-C,

I dont think Align.coerce can handle integer values, so I added conditions for the 3 possible types integer, string and Align.

File: test/outline/test_outline.py (hunk 3 above)
Comment (visheshdvivedi) [reply test/outline/test_outline.py:68]: thanks

**Comment by Lucas-C** [comment]
Hi @visheshdvivedi

Just for your information, several occurences of `title_style` were renamed into `text_style` in commit https://github.com/py-pdf/fpdf2/commit/7b3f1bdd23026b7d68a38699c7fe1d6977080c3f last week.

I think this caused minor issues in your PR when you rebased your branch:
```
fpdf/fpdf.py:5153:31: E0602: Undefined variable 'title_style' (undefined-variable)
fpdf/fpdf.py:5169:11: E0602: Undefined variable 'text_style' (undefined-variable)
fpdf/fpdf.py:5169:26: E0602: Undefined variable 'text_style' (undefined-variable)
fpdf/fpdf.py:5170:20: E0602: Undefined variable 'text_style' (undefined-variable)
```

**Comment by Lucas-C** [comment]
Hi @visheshdvivedi 🙂

Just to get a quick update: do you still want to implement this feature?

**Comment by visheshdvivedi** [comment]
Hi @Lucas-C ,

Yes, very much.

Aplogies for the delay on my side, have been occupied for sometime in personal tasks. But I will try to put more time on this and complete it.

**Comment by Lucas-C** [comment]
> Aplogies for the delay on my side, have been occupied for sometime in personal tasks. But I will try to put more time on this and complete it.

No worries 👍

Just please keep us informed if ever you realized you won't be able to finish this, and the maintainers will take over 🙂

**Comment by visheshdvivedi** [comment]
Hi @Lucas-C ,

I have completed the code and written the tests, but having issues with passing all tests. Have been stuck here for sometime now. Would you be able to assist ?

Thanks for your support

**Comment by andersonhc** [comment]
@visheshdvivedi 
After some debugging I found out all float l_margin were defaulting to 0 due to the insinstance check. 
I rebased your branch and pushed a fix. 
Please review and let us know if it's ready or if you have more work to do on this PR. 

Thanks

**Comment by Lucas-C** [comment]
> After some debugging I found out all float l_margin were defaulting to 0 due to the insinstance check.
> I rebased your branch and pushed a fix.

Thanks a lot @andersonhc, well done!

I added a commit to introduce extra checks when unsupported `text_style.l_margin` values are provided to `.use_text_style()` or `TableOfContents()`.

I also added one last commit to support `Align` values provided as `l_margin` in HTML tags styles.

I'm going to merge this now in order to include this nice feature in the next release 🙂

Thank you @visheshdvivedi for your contribution 👍

**Comment by Lucas-C** [comment]
@allcontributors please add @visheshdvivedi for code

**Comment by visheshdvivedi** [comment]
Hi @Lucas-C  and @andersonhc,

Thanks you so much for your help and support. This is kinda like my first proper open source contribution and I wouldn't have been able to make it without your help.

If you ever feel like any new feature or bug request is required, please message me I will gladly help with the same.

**Comment by Lucas-C** [comment]
> Thanks you so much for your help and support. This is kinda like my first proper open source contribution and I wouldn't have been able to make it without your help.

You are very welcome 🙂

By the way, your contribution has been included in the new release of `fpdf2`!
https://github.com/py-pdf/fpdf2/releases/tag/2.8.2

**Pull request merged**